File: cli/src/etos_client/etos/v0/test_results/test_results.py
# Copyright Axis Communications AB.
#
# For a full list of individual contributors, please see the commit history.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""ETOS test result handler."""
import logging
from typing import Optional

from ..events.events import Events, TestSuite

# pylint:disable=too-few-public-methods


class TestResults:
    """Get test results from an ETOS testrun."""

    logger = logging.getLogger(__name__)

    def __has_failed(self, test_suites: list[TestSuite]) -> bool:
        """Check if any test suite has failed in a list of test suites."""
        for test_suite in test_suites:
            outcome = test_suite.finished["data"]["testSuiteOutcome"]
            if outcome["verdict"] != "PASSED":
                return True
        return False

    def __fail_messages(self, test_suites: list[TestSuite]) -> list[str]:
        """Build a fail message from main suites errors."""
        messages = []
        for test_suite in test_suites:
            outcome = test_suite.finished["data"]["testSuiteOutcome"]
            if outcome["conclusion"] != "SUCCESSFUL":
                messages.append(f'{test_suite.started["data"]["name"]}: {outcome["description"]}')
        return messages

    def __test_result(self, test_suites: list[TestSuite]) -> tuple[bool, str]:
        """Build test results based on events retrieved."""
        if not self.__has_failed(test_suites):
            return True, "Test suite finished successfully."
        messages = self.__fail_messages(test_suites)
        if len(messages) == 1:
            return False, messages[0]
        if messages:
            for message in messages[:-1]:
                self.logger.error(message)
            return False, messages[-1]
        return False, "Test case failures during test suite execution"

    def get_results(self, events: Events) -> tuple[Optional[bool], Optional[str]]:
        """Get results from an ETOS testrun."""
        if not events.tercc:
            return None, None
        if not events.activity.triggered:
            return None, None
        if events.activity.canceled:
            return None, None
        if not events.main_suites:
            return None, None
        for test_suite in events.main_suites:
            if not test_suite.finished:
                return None, None
        return self.__test_result(events.main_suites)

File: cli-go/pkg/client/v0/test_results/test_results.go
package test_results

import (
	"fmt"

	"github.com/eiffel-community/eiffelevents-sdk-go/editions/orizaba"
	"github.com/fredjn/etos/cli-go/pkg/client/v0/events"
)

// TestResults handles test results from an ETOS test run
type TestResults struct {
	events *events.Events
}

// NewTestResults creates a new test results handler
func NewTestResults(events *events.Events) *TestResults {
	return &TestResults{
		events: events,
	}
}

// hasFailed checks if any test suite has failed
func (t *TestResults) hasFailed() bool {
	for _, suite := range t.events.MainSuites {
		if suite.Finished == nil {
			continue
		}
		finished := suite.Finished.Get().(*orizaba.TestSuiteFinished)
		if finished.Data.Outcome.Conclusion != "SUCCESSFUL" {
			return true
		}
	}
	return false
}

// failMessages builds fail messages from main suites errors
func (t *TestResults) failMessages() []string {
	var messages []string
	for _, suite := range t.events.MainSuites {
		if suite.Finished == nil {
			continue
		}
		finished := suite.Finished.Get().(*orizaba.TestSuiteFinished)
		if finished.Data.Outcome.Conclusion != "SUCCESSFUL" {
			started := suite.Started.Get().(*orizaba.TestSuiteStarted)
			messages = append(messages, fmt.Sprintf("%s: %s", started.Data.Name, finished.Data.Outcome.Description))
		}
	}
	return messages
}

// GetResults gets the results from an ETOS test run
func (t *TestResults) GetResults() (bool, string, error) {
	// Check if we have all required events
	if t.events.TERCC == nil {
		return false, "", fmt.Errorf("no TERCC event found")
	}
	if t.events.Activity == nil || t.events.Activity.Triggered == nil {
		return false, "", fmt.Errorf("no activity triggered event found")
	}
	if t.events.Activity.Canceled != nil {
		return false, "", fmt.Errorf("activity was canceled")
	}
	if len(t.events.MainSuites) == 0 {
		return false, "", fmt.Errorf("no main test suites found")
	}
	for _, suite := range t.events.MainSuites {
		if suite.Finished == nil {
			started := suite.Started.Get().(*orizaba.TestSuiteStarted)
			return false, "", fmt.Errorf("test suite %s not finished", started.Meta.ID)
		}
	}

	// Check if any test suite failed
	if !t.hasFailed() {
		return true, "Test suite finished successfully.", nil
	}

	// Get fail messages
	messages := t.failMessages()
	if len(messages) == 1 {
		return false, messages[0], nil
	}
	if len(messages) > 0 {
		return false, messages[len(messages)-1], nil
	}

	return false, "Test case failures during test suite execution", nil
} 